services/server/config/config.go

```diff
@@ -14,6 +14,12 @@
    limitations under the License.
 */
 
+// config is the global configuration for containerd
+//
+// Version History
+// 1: Deprecated and removed in containerd 2.0
+// 2: Uses fully qualified plugin names
+// 3: Added support for migration and warning on unknown fields
```

Comment:
v3 just seems to update implementation details.
Can we just retain v2 as the format version?

Reply:
Would you suggest we add another version here? We did that with database to differentiate "schema version" from the data version.

Reply:
Can't we just keep it a single "version: 2" ?

Reply:
We need some way to know a migration has been completed, otherwise deprecated fields will always be looked for.

Reply:
Maybe just add Annotation map[string]string to store the migration state?

Reply:
I think its best we keep the versioning like this, the other options aren't nearly as clean or easy to handle the migration per version. Realistically these versions won't continue bump very fast since we don't introduce new plugins very often. At most this once per minor release. Keeping the versioning like this allows us to move forward without breaking older configurations, support explicit migration, or allow users to switch between older version of containerd without migrating. 

I'm also looking into whether we can just un-deprecate version 1 and use a similar migration. This would be cleaner to users than having an error. I could include this here https://github.com/containerd/containerd/commit/5fd2bed60d0ef80e8c635d7b242e5aafe407c7ec (still needs docs updates)